### ISSUE TITLE
Highlights refactoring broke "highlight.ERROR/WARNING"

### ISSUE THREAD

**Issue by cixtor** [SublimeLinter/SublimeLinter#915]
[This refactoring](https://github.com/SublimeLinter/SublimeLinter/commit/f3a57dc) of the `highlight` module broken the [SublimeLinter-annotations](https://github.com/SublimeLinter/SublimeLinter-annotations) linter. I was going to send a pull-request to the other repository but figured it's better to report the issue here since the code is still under development _(for version 4.0.0)_.

```
reloading plugin SublimeLinter-annotations.linter
Traceback (most recent call last):
  File "/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 109, in reload_plugin
    m = importlib.import_module(modulename)
  File "./python3.3/importlib/__init__.py", line 90, in import_module
  File "<frozen importlib._bootstrap>", line 1584, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1565, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1532, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 584, in _check_name_wrapper
  File "<frozen importlib._bootstrap>", line 1022, in load_module
  File "<frozen importlib._bootstrap>", line 1003, in load_module
  File "<frozen importlib._bootstrap>", line 560, in module_for_loader_wrapper
  File "<frozen importlib._bootstrap>", line 868, in _load_module
  File "<frozen importlib._bootstrap>", line 313, in _call_with_frames_removed
  File "/Users/root/Library/Application Support/Sublime Text 3/Packages/SublimeLinter-annotations/linter.py", line 15, in <module>
    from SublimeLinter.lint import highlight, Linter
ImportError: cannot import name highlight
```

**Comment by braver** [comment]
`highlight.ERROR/WARNING` is now `const.ERROR/WARNING`, breaking:

- [ ] SublimeLinter-contrib-golint
- [ ] SublimeLinter-contrib-gometalinter
- [ ] SublimeLinter-contrib-gosimple
- [ ] SublimeLinter-contrib-mlint
- [ ] SublimeLinter-contrib-mypy
- [ ] SublimeLinter-contrib-protoc-gen-lint
- [ ] SublimeLinter-contrib-pug-lint
- [ ] SublimeLinter-contrib-write-good
- [ ] SublimeLinter-foodcritic
- [ ] SublimeLinter-jsdebugmarkers
- [ ] SublimeLinter-annotations
- [ ] SublimeLinter-phpmd
- [ ] SublimeLinter-pydocstyle
- [x] SublimeLinter-rst (deprecated)

**Comment by braver** [comment]
related: 697c7c8c10caed81b29c69d17b6db7bdf4a355ca

**Comment by braver** [comment]
Example fix: https://github.com/SublimeLinter/SublimeLinter-annotations/commit/b2862845d19e04f861d161aceddf4775a1bb6c63

**Comment by kaste** [comment]
Didn't know that `highlight` was a public module in SL3. (And it is completely ... to import ERROR and WARNING from highlight.) 

But we should release a patch here.

**Comment by braver** [comment]
You live, you learn

**Comment by gerardroche** [comment]
I just sent a [pr fix](https://github.com/fredcallaway/SublimeLinter-contrib-mypy/pull/21) for mypy. You can cross that one off the list.

**Comment by cixtor** [comment]
I created a pull-request to fix SublimeLinter-contrib-gometalinter [1]

[1] https://github.com/alecthomas/SublimeLinter-contrib-gometalinter/pull/12

**Comment by braver** [comment]
We may move ERROR/WARNING again before the night is done, watch this space!

**Comment by cixtor** [comment]
I have created a pull-request to fix these linters:

- [x] SublimeLinter-contrib-golint — https://github.com/sirreal/SublimeLinter-contrib-golint/pull/6
- [x] SublimeLinter-contrib-gometalinter — https://github.com/alecthomas/SublimeLinter-contrib-gometalinter/pull/12
- [ ] SublimeLinter-contrib-gosimple

I will watch out for the changes of ERROR/WARNING and update the pull-requests.

**Comment by cixtor** [comment]
I have created a pull-request to fix these linters:

- [x] SublimeLinter-contrib-golint — https://github.com/sirreal/SublimeLinter-contrib-golint/pull/6
- [x] SublimeLinter-contrib-gometalinter — https://github.com/alecthomas/SublimeLinter-contrib-gometalinter/pull/12
- [ ] SublimeLinter-contrib-gosimple

I will watch out for the changes of ERROR/WARNING and update the pull-requests.

**Comment by braver** [comment]
Dudes, totally appreciate the help but wait a second to see if we merge #929 before you have to go and do it again.

**Comment by braver** [comment]
@kaste I slept on it. Should do more of that. New proposal:

- starting from SL4 plugins should import ERROR/WARNING from `const`
- I go and update plugins to do so, with the help of @cixtor (dude, your user name is annoying as hell by the way)
- meanwhile your hot fix is already out there so no plugins are actually broken right now
- when we're done and every plugin is prepared we can revert d3ac90c94f0bfe6d5d78be3aae44faa4ed3bb049

**Comment by kaste** [comment]
I think we should do 

```
from SublimeLinter import Linter, ERROR, WARNING, STDOUT
```

Everything a normal plugin could use.

I don't like `const.py` so I don't necessarily want to expose that.

```
PROTECTED_REGIONS_KEY = "sublime_linter.protected_regions"
STATUS_KEY = "sublime_linter_status"
STATUS_BUSY_KEY = "sublime_linter_status_busy"

WARNING = "warning"
ERROR = "error"
WARN_ERR = (WARNING, ERROR)
INBUILT_ICONS = ("circle", "dot", "bookmark", "none")
```

Everything in there except ERROR, WARNING should go to the appropriate view modules. We don't really share these consts. E.g. STATUS_KEY is private to our status_bar_view. INBUILT_ICONS is either private to highlights or style.py, don't remember. WARN_ERR is basically unused.

**Comment by kaste** [comment]
@cixtor Your username looks like this 
![image](https://user-images.githubusercontent.com/8558/35730110-21aec062-0811-11e8-818a-b3dbdb9d79ec.png)

**Comment by braver** [comment]
Ok, I can totally get behind that. Was #929 on the right track then?

**Comment by kaste** [comment]
Well, if you rewrite every line in that. Kinda sorta 😁

**Comment by braver** [comment]
"Kinda sorta" I can work with. Do you trust me to have another go at it or would you rather do it yourself? 😁

**Comment by gerardroche** [comment]
I'm haven't read all the threads related to this issue, so if I'm rehashing something, sorry.

Why not create a explicit public api module for plugins? For example `lint/plugin`. Then plugins import *only* things from that module. If plugins need to use something else, they should open an issue to expose it as a public api.

By way of example, the `mypy` plugin [imports are like this](https://github.com/gerardroche/SublimeLinter-contrib-mypy/blob/cd87f219a2bad3797a28d9a2eb387d52d2b7eda7/linter.py): 

```
from SublimeLinter.lint import const
from SublimeLinter.lint import persist
from SublimeLinter.lint import PythonLinter

# and then accesses the above like:

class Mypy(PythonLinter):
# ...
    default type = const.WARNING
# ...
        persist.printf("mypy: Unable to delete '{}' while cleaning up temporary directory".format(path))
```

Instead it would either import like this:

```
from SublimeLinter.lint.plugin import const
from SublimeLinter.lint.plugin import persist
from SublimeLinter.lint.plugin import PythonLinter
```

Or create a cleaner API:

```
from SublimeLinter.lint import plugin

# and then access things something like:

class Mypy(plugin.PythonLinter):
# ...
    default type = plugin.DEFAULT_TYPE_WARNING
# ...
        plugin.log_info("mypy: Unable to delete '{}' while cleaning up temporary directory".format(path))
```

This will make it easier to reorganise things in the core while keeping the plugin api compatible.

**Comment by braver** [comment]
Well, basically everything in lint is available for plugins to extend or modify. The docs and templates should guide authors in the right direction. We can’t really go back and redesign it because there are over 150 plugins out there. 

So our approach is a bit situational. We’re careful about breakage, sometimes we go out and modify the plugins before a change, sometimes we have to fix something afterwards.

**Comment by kaste** [comment]
@gerardroche Why ``from SublimeLinter.lint.plugin import WARNING`` instead of ``from SublimeLinter import WARNING`` ?

**Comment by gerardroche** [comment]
@kaste I guess it's just to make it more explicit that it is the plugin endpoint. Plus ST has issues with importing from the base package or root package modules e.g. it can cause double loading issues. So `from SublimeLinter.plugin import ...` which would be better is not really an option, unless "plugin" in this case is package and not a module, although ST probably has issue with that too.  You could go with `from SublimeLinter.plugin.api import ...`. It's really just about finding the best naming convention that ST doesn't have problems with e.g. not so older versions of ST also has issues with anything named "plugins".

**Comment by kaste** [comment]
@gerardroche Yeahhhh, okay, it loads them multiple times, I see that, but ... help me ... an `__init__.py` has no side effects. It just imports stuff from down and exposes it. It will not reload our probably (hopefully not) stateful modules in the *lint* package (I checked that.). So we *could* not care? No?

(Problem: Stupid or naive as we are, we expose a VERSION on `SublimeLinter.__init__.py`)

**Comment by alecthomas** [comment]
So what's the verdict here? Should I merge https://github.com/alecthomas/SublimeLinter-contrib-gometalinter/pull/12? Will that also work with SublimeLinter v3.x?

**Comment by braver** [comment]
@alecthomas there is no need for action right now, all plugins should work with current releases of SL 3 and 4. We’re trying to figure out if we have a problem with double loading before we move forward with #932

**Comment by kaste** [comment]
No, our *views* are completely safe. We just erred with `SublimeLinter.VERSION` which must be `SublimeLinter.lint.VERSION`

The rule here is *views* only **react** upon events. It is forbidden that they import each other. And it is forbidden that the backend imports (acts as a controller of) a view. (Essentially, views don't have controllers. Bc that's soooo old fashioned OOP style.)

**Comment by braver** [comment]
@cixtor @alecthomas We are backwards compatible right now, so there is no hurry, but we will eventually drop the compatibility. The ERROR/WARNING constants are now at `const` and will remain there. The correct way to check for the SL version is now `getattr(SublimeLinter.lint, 'VERSION', 3)`.

@cixtor feel free to open a PR at SublimeLinter-contrib-gosimple, but make sure to include a switch on that version to maintain compatibility.

**Comment by gerardroche** [comment]
I've had a few different loading issues, including issues using `__init__.py`. Perhaps some of these issues have been fixed. I'm not convinced. I know that some issues have been fixed, like the issue with things named "plugins", but even that issue was only fixed in a recent build so it still affects users who are on the ST3 general release. I don't know all the specifics of all the related issues. I've just learned to avoid importing from the root package, importing package root modules, or using a root package `__init__.py`, because it's sure to come back a bite you.
